docs/design/script-runner.md

```diff
@@ -0,0 +1,102 @@
+    // Execution results will not be updated if not specified.
+    // +optional
+    OperationResultKey *string `json:"operationResultKey,omitempty"`
+}
+```
+
+The `ScriptRunner` processor is designed for users to define how should KubeDiag run their scripts. It contains the following fields:
+
+* `.spec.processor.scriptRunner.script`: The content of shell script.
+* `.spec.processor.scriptRunner.argKeys`: ArgKeys contains a slice of keys in parameters or operationResults. The script arguments are generated from specified key value pairs. No argument will be passed to the script if not specified.
+* `.spec.processor.scriptRunner.operationResultKey`: OperationResultKey is the prefix of keys to store script stdout, stderr or error message in operationResults. Execution results will not be updated if not specified.
+
+### Managing Scripts on Node
+
+KubeDiag stores all scripts under the agent data root directory. The operation controller will create, update or delete an executable file if an `Operation` of script runner type is reconciled. The following steps will happen during reconciliation:
+
+* Operation controller in agent reconciles the created operation and check if `.spec.processor.scriptRunner` is nil.
+* If the operation is not a script runner, skip synchronizing the object.
+* If the operation is a script runner and event type is create, the agent will create a file under data root directory, which defaults to `/var/lib/kubediag/scripts`.
+  * Change file mode of the script to executable.
+* If the operation is a script runner and event type is update, the agent will remove the older file with operation name and create a new file under data root directory.
+  * Change file mode of the script to executable.
+* If the operation is a script runner and event type is delete, the agent will remove the file with operation name under data root directory.
+
+### Passing Arguments to a Script
+
+An `Operation` is defined before any execution of script. The `Diagnosis` resource stores details of a pipeline run, which includes how can we pass arguments to a script defined in `Operation`.
+
+To pass any arguments to a script when running a diagnosis, `.spec.parameters` of `Diagnosis` should be specified on creation. A scripts can also take results from precedent operations as arguments if any component in `.spec.processor.scriptRunner.argKeys` of `Operation` matches any keys in `.status.operationResults` of `Diagnosis`. Arguments will be constructed from `.spec.parameters` and `.status.operationResults` just in time before script run. For example, a `Diagnosis` might contain the following JSON object in `.spec.parameters` and `.status.operationResults`:
+
+```json
+{
+    "key1": "value1",
+    "key2": "value2",
+    "key3": "value3",
+    "key4": "value4"
+}
+```
+
+If `.spec.processor.scriptRunner.argKeys` of `Operation` is `["key1", "key4"]`, then `value1` and `value4` are passed as arguments to the script.
```

Comment:
I think `spec.processor.scriptRunner.argKeys` is not required for this field. User can directly pass the script args of an operation in the form of a string directly in `diagnosis.spec.parameter`, so the design will be much simpler

Reply:
This feature is design for managing scripts written before KubeDiag is introduced in cluster. The author of a script may not be aware of such mechanism for context propagation. Specifying `.spec.processor.scriptRunner.argKeys` explicitly would help users to manage their legacy scripts with our framework. Users are also capable to pass results of precedent operations to a script runner during diagnosis.